gemini/src/main/java/com/techempower/gemini/path/AnnotationDispatcher.java

```diff
@@ -0,0 +1,550 @@
+    private Reflections      reflections            = null;
+
+    public AnnotationDispatcher(GeminiApplication application)
+    {
+        app = application;
+        handlers = new HashMap<>();
+        exceptionHandlers = new ExceptionHandler[]{};
+        prehandlers       = new Prehandler[]{};
+        listeners         = new DispatchListener[]{};
+
+        if (exceptionHandlers.length == 0)
+        {
+            throw new IllegalArgumentException("PathDispatcher must be configured with at least one ExceptionHandler.");
+        }
+
+        startReflectionsThread();
+    }
+
+    private void startReflectionsThread()
+    {
+        // Start constructing Reflections on a new thread since it takes a
+        // bit of time.
+        preinitializationTasks.submit(new Runnable() {
+            @Override
+            public void run() {
+                try
+                {
+                    reflections = PackageClassLoader.getReflectionClassLoader(app);
+                }
+                catch (Exception exc)
+                {
+                    // todo
+//                    log.log("Exception while instantiating Reflections component.", exc);
+                }
+            }
+        });
+    }
+
+    public void initialize() {
+        // Wait for pre-initialization tasks to complete.
+        try
+        {
+//            log.log("Completing preinitialization tasks.");
+            preinitializationTasks.shutdown();
+//            log.log("Awaiting termination of preinitialization tasks.");
+            preinitializationTasks.awaitTermination(5L, TimeUnit.MINUTES);
+//            log.log("Preinitialization tasks complete.");
+//            log.log("Reflections component: " + reflections);
+        }
+        catch (InterruptedException iexc)
+        {
+//            log.log("Preinitialization interrupted.", iexc);
+        }
+
+        // Throw an exception if Reflections is not ready.
+        if (reflections == null)
+        {
+            throw new ConfigurationError("Reflections not ready; application cannot start.");
+        }
+
+        register();
+    }
+
+    private void register() {
+//        log.log("Registering annotated entities, relations, and type adapters.");
+        try {
+            final ExecutorService service = Executors.newFixedThreadPool(1);
+
+            // @Path-annotated classes.
+            service.submit(new Runnable() {
+                @Override
+                public void run() {
+                    for (Class<?> clazz : reflections.getTypesAnnotatedWith(Path.class)) {
+                        final Path annotation = clazz.getAnnotation(Path.class);
+
+                        try {
+                            handlers.put(annotation.value(),
+                                    new AnnotationHandler(annotation.value(),
+                                            clazz.getDeclaredConstructor().newInstance()));

```

Comment:
If you are going to have classpath scanning for the automatic instantiation of handlers, I think you should:
- have a means of opting-out of this with an annotation
- skip instantiating handlers that are already registered (meaning there should be a way for a dev to manually register a handler with the dispatcher before getting to this point)
- provide some form of dependency injection using Google Guice or something, though if you do the above point then in theory this isn't necessary. A dev would be able to use the dependency injection framework of their choosing to init handlers ahead of time, which would in turn omit them from the classpath-scanning-initialization per that previous point. I have no experience whatsoever with dependency injection though. I might look into this later on.

Reply:
Why? If your `constructDispatcher` instantiates a new `AnnotationDispatcher` you *want* to scan using reflection... right?

Reply:
re-my first point: Ignore this point. While I think it could still be useful to completely disable a handler that you don't want accessible without removing it from the repo for legacy reasons, etc etc, it's whatever.

re-your comment: Your approach assumes that the handler's constructor takes no parameters. It's very possible for that not to be true. Requiring a zero-args constructor prevents the handlers from being provided a reference to a database service (for example) unless you're using `getInstance()` methods (very bad for tests) or providing the necessary instances to each handler after initialization. That's why the other option is dependency injection, but unless you're familiar with that or want to become familiar with that, having the alternative of manually registering handlers that you initialize yourself seems like a simple enough solution.